src/webgpu/api/operation/memory_sync/buffer/write_after_write_in_buffer.spec.ts

```diff
@@ -0,0 +1,113 @@
+
+import { BufferSyncTest } from './buffer_sync_test.js';
+
+export const g = makeTestGroup(BufferSyncTest);
+
+g.test('write_after_write')
+  .desc(
+    `Test write-after-write operations. The first write will write 1 into a storage buffer.
+    The second write will write 2 into the same storage buffer. So, expected data in buffer is 2.
+    The two writes can be in the same command buffer, or separate command buffers, or separate
+    submits, or separate queues. Each write operation can be done via render, compute, copy,
+    writeBuffer, map write, etc. If the write operation is done by a render pass, it may use bundle.`
+  )
+  .params(
+    params()
+      // TODO (yunchao.he@intel.com): add multi-queue.
+      .combine(
+        poptions('writeScopes', ['sameCmdbuf', 'separateCmdbufs', 'separateSubmits'] as const)
+      )
+      .combine(
+        poptions('firstWriteOp', [
+          'render',
+          'compute',
+          'b2bCopy',
+          't2bCopy',
+          'writeBuffer',
+        ] as const)
+      )
+      .combine(
+        poptions('secondWriteOp', [
+          'render',
+          'compute',
+          'b2bCopy',
+          't2bCopy',
+          'writeBuffer',
+        ] as const)
+      )
+      .combine(pbool('firstWriteInBundle'))
+      .combine(pbool('secondWriteInBundle'))
+      .unless(
+        p =>
+          (p.firstWriteInBundle && p.firstWriteOp !== 'render') ||
+          (p.secondWriteInBundle && p.secondWriteOp !== 'render') ||
+          ((p.firstWriteOp === 'writeBuffer' || p.secondWriteOp === 'writeBuffer') &&
+            p.writeScopes !== 'separateSubmits')
+      )
+  )
+  .fn(async t => {
+    const {
+      writeScopes,
+      firstWriteOp,
+      secondWriteOp,
+      firstWriteInBundle,
+      secondWriteInBundle,
+    } = t.params;
+
+    const buffer = await t.createBufferWithValue(0);
+
+    const writeInBundle = [firstWriteInBundle, secondWriteInBundle];
+    const writeOp = [firstWriteOp, secondWriteOp];
+    switch (writeScopes) {
+      case 'sameCmdbuf': {
+        const encoder = t.device.createCommandEncoder();
+        for (let i = 0; i < 2; i++) {
+          await t.issueWriteOp(writeOp[i], writeInBundle[i], buffer, i + 1, encoder);
+        }
+        t.device.defaultQueue.submit([encoder.finish()]);
+        break;
+      }
+      case 'separateCmdbufs': {
+        const command_buffers: GPUCommandBuffer[] = [];
+        for (let i = 0; i < 2; i++) {
+          command_buffers.push(
+            await t.createCommandBufferAndIssueWriteOp(writeOp[i], writeInBundle[i], buffer, i + 1)
+          );
+        }
+        t.device.defaultQueue.submit(command_buffers);
+        break;
+      }
+      case 'separateSubmits': {
+        for (let i = 0; i < 2; i++) {
+          await t.createQueueSubmitsAndIssueWriteOp(writeOp[i], writeInBundle[i], buffer, i + 1);
+        }
+        break;
+      }
+      default:
+        assert(true);
```

Comment:
I think you meant `assert(false)` though `unreachable()` would also work. This code will disappear with the suggestion above though.

Reply:
Ack. Done.